system/lib/pthread/pthread_self.s

```diff
@@ -0,0 +1,39 @@
+.globaltype thread_id, i32
+thread_id:
+
+.globaltype is_main_thread, i32
+is_main_thread:
+
+.globaltype is_runtime_thread, i32
+is_runtime_thread:
+
+.globl pthread_self
+pthread_self:
```

Comment:
Wait, so any function that is defined in `.s` file gets a `_` prefixed to it? So `pthread_self` here becomes `_pthread_self`?

And any functions defined in C/C++ files do not get a `_` prefix?

That is actually pretty backwards: in native compilers all assembly code is unprefixed, and all C/C++ code is prefixed with `_`. Is there a specific reason why this was flipped for Wasm? That seems a bit arbitrary.

Reply:
The `_` prefix is added to all native symbols by the emscripten JS code.   We do this purely for legacy/asm.js compatability.

The actual compiler, llvm, does not emit the `_` mangling for any symbols.     I think maybe the `_` mangling is some kind of mac or windows convention.  It certainly doesn't exist on windows.

From the linker's POV there is no way to way to the difference between a C symbol and an assembler symbols.  They are all just "native" sybmols and we mangle all of them when we import them into JS.

One day we should completely remove the mangling...

Reply:
I created an issue to remove this mangling completely one day: https://github.com/emscripten-core/emscripten/issues/12844